applications/nrf_desktop/src/util/config_channel_transport.c

```diff
@@ -0,0 +1,272 @@
+	0, 		       /* Event ID */
+	CONFIG_STATUS_PENDING, /* Status */
+	0,		       /* Event data length */
+};
+
+static int frame_length_check(const size_t length)
+{
+	const size_t min_size = TRANSPORT_HEADER_SIZE;
+	const size_t max_size = REPORT_SIZE_USER_CONFIG;
+
+	BUILD_ASSERT(min_size < max_size);
+
+	if ((length < min_size) || (length > max_size)) {
+		LOG_WRN("Unsupported report length %zu", length);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int data_len_check(const size_t event_data_len)
+{
+	const size_t min_size = TRANSPORT_HEADER_SIZE;
+	const size_t max_size = REPORT_SIZE_USER_CONFIG;
+
+	BUILD_ASSERT(min_size < max_size);
+
+	if (event_data_len > max_size - min_size) {
+		LOG_WRN("Unsupported event data length %" PRIu8,
+			event_data_len);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int config_channel_report_parse(const uint8_t *buffer, size_t length,
+				struct config_event *event)
+{
+	int err = frame_length_check(length);
+
+	if (err) {
+		return err;
+	}
+
+	size_t pos = 0;
+
+	event->recipient = sys_get_le16(&buffer[pos]);
+	pos += sizeof(event->recipient);
+
+	event->event_id = buffer[pos];
+	pos += sizeof(event->event_id);
+
+	event->status = buffer[pos];
+	pos += sizeof(event->status);
+
+	uint8_t config_data_len = buffer[pos];
+	pos += sizeof(uint8_t);
+
+	__ASSERT_NO_MSG(pos == TRANSPORT_HEADER_SIZE);
+
+	err = data_len_check(config_data_len);
+	if (err) {
+		return err;
+	}
+
+	if ((length - pos) < config_data_len) {
+		LOG_ERR("Not enough data in buffer");
+		return -ENOTSUP;
+	}
+
+	__ASSERT_NO_MSG(event->dyndata.size >= config_data_len);
+	event->dyndata.size = config_data_len;
+
+	memcpy(event->dyndata.data, &buffer[pos], config_data_len);
+	pos += config_data_len;
+
+	return pos;
+}
+
+int config_channel_report_fill(uint8_t *buffer, const size_t length,
+			       const struct config_event *event)
+{
+	int err = frame_length_check(length);
+	if (!err) {
+		err = data_len_check(event->dyndata.size);
+	}
+
+	if (err) {
+		return err;
+	}
+
+	size_t pos = 0;
+
+	sys_put_le16(event->recipient, &buffer[pos]);
+	pos += sizeof(event->recipient);
+
+	buffer[pos] = event->event_id;
+	pos += sizeof(event->event_id);
+
+	buffer[pos] = event->status;
+	pos += sizeof(event->status);
+
+	buffer[pos] = (uint8_t)event->dyndata.size;
+	pos += sizeof(uint8_t); /* Data length is encoded using one byte. */
+
+	__ASSERT_NO_MSG(pos == TRANSPORT_HEADER_SIZE);
+	__ASSERT_NO_MSG(pos + event->dyndata.size <= length);
+
+	if (event->dyndata.size > 0) {
+		memcpy(&buffer[pos], event->dyndata.data, event->dyndata.size);
+		pos += event->dyndata.size;
+	}
+
+	return pos;
+}
+
+static void drop_transactions(struct config_channel_transport *transport)
+{
+	/* Transport ID is changed, responses for submitted requests will
+	 * be dropped.
+	 */
+	uint8_t *new_ptr = transport->transport_id;
+
+	new_ptr++;
+	/* Make sure we will not overlap with another transport. */
+	if ((new_ptr - (uint8_t *)transport) >= sizeof(*transport)) {
```

Comment:
now that is an abuse I haven't seen in a long time... long time...

Pointer arithmetic is dangerous as if used incorrectly it can lead to undefined behavior and dragons. But apart from that... why not simply use integers?

Reply:
I will switch to using uint16_t here